Rewrite of the codemod

danielberndt opened the pull request:
Hi there!

I eventually found some time to dig into AST transformations, and I had the desire to dig deep 😅 

This PR is more or less a complete rewrite, supporting all the old-features, and including many new ones. Just check the `__tests__/__fixtures__` folder for some examples.

It also allows you to run tests via `npm install && npm test -- --watch` for a fast feedback loop while developing. 

I haven't tested it on a real code base yet, so consider this more of an WIP...

Here's what this PR should fix from my [list](https://github.com/paypal/glamorous/issues/419#issuecomment-407113266):

- [x] wont work with glamorous.macro (should be easy to fix)
- [x] won't work if you do imports like `import g from "glamorous"` the codemod expects import glamorous from "glamorous"
- [x] you manually need to convert code like `<glamorous.div marginTop={5}/>` to `<div css={{marginTop: 5}}/>`
- [ ] no automatic translation from `filterProps/forwardProps` to `shouldForwardProps` (the latter expects a function), most other options (e.g. withProps) are also not directly translatable
- [ ] `forwardProps: ["innerRef"]` does not work with emotion
- [ ] global styles and keyframes can't be automatically translated yet either

danielberndt commented:
Here's an updated version of the Explorer: https://astexplorer.net/#/gist/7bc4771564a12c9f93c4904b3934aa1c/b8ce61a62c97c913866b42426b24762c4c1d0c83

TejasQ commented:
Dude. Thank you so much! That was incredible! I was actually planning to set aside time to write up a part of this for you to have a read and possibly learn from. Looks like you did amazing stuff already!

Can I participate in this PR and help with the remaining tasks?

Thanks again for the PR!

danielberndt commented:
I think I'd like to try continuing with global styles and keyframes. But if you want you can maybe look into improving which properties are actual css properties and which are not [here](https://github.com/danielberndt/babel-plugin-glamorous-to-emotion/blob/a2ded59032180161f77d96724d8878ad7d47a637/index.js#L56-L57).

e.g `<glamorous.Button marginTop={5} disabled={true}/>` should not become `<button css={{marginTop: 5, disabled: true}}/>`. I guess that glamorous is using some external library for getting a list of all valid css attributes. So maybe we can use this here as well.

Also one could attempt to fail the codemod if it encounters `filterProps/forwardProps` and add some helpful message of which file at which location these are used and what method to use in emotion to replace them. (i.e. asking the user to change their code to `const Comp = glamorous(MyComp, {shouldForwardProps: p => ...})({...style})` so that the resulting conversion to emotion will then work... or maybe try to automatically translate it, if it's just a simple list within `filterProps`

TejasQ commented:
@danielberndt I'll look at that. Thanks for your contributions so far!

TejasQ commented on index.js:
```diff
@@ -13,133 +13,167 @@
+module.exports = function(babel) {
+  const {types: t} = babel;
+
+  const fixContentProp = args => {
```
What is the expected value of `args`?

Can we add a bit of documentation here so it's easier for people to understand? I'd like this project to be a bit educational as well: it helped you, so it ought to help others interested in AST transformation.

danielberndt replied:
Yes you're, right, I'll do a more descriptive pass soon :)

TejasQ commented on index.js:
```diff
@@ -13,133 +13,167 @@
+    const newAttrs = attrs.filter(attr => {
+      const {value, name, type} = attr;
+      if (type === "JSXSpreadAttribute") return true;
+      if (name.name === "css") {
```
I think we can do better than `name.name` here in order to allow developers not as experienced with this stuff to understand what's happening and more clearly describe our intent.

danielberndt replied:
I didn't come up with `attr.name.name`. That's part of the AST API.

TejasQ replied:
I know that, but I'm suggesting aliasing the destructured name on line 51 in order to provide an identifier that is a little more understandable.

TejasQ commented on index.js:
```diff
@@ -13,133 +13,167 @@
-                );
+        appendToCss.push({
+          name: t.identifier(name.name),
+          value: value.type === "JSXExpressionContainer" ? value.expression : value,
```
Would the `t.isJSXExpressionContainer` function work for this validation instead? I think it's more future proof because it could handle more imperative validations if needed in the future.

danielberndt replied:
This probably applies to `type === "XXX"` checks, of which I did quite a lot.

TejasQ commented on index.js:
```diff
@@ -13,133 +13,167 @@
+    if (appendToCss.length > 0) {
+      if (!cssAttr) {
+        cssAttr = t.jsxAttribute(
+          t.jsxIdentifier("css"),
```
Afaik, `<span css={}` is invalid in intrinsic JSX elements. The valid alternative is `<span style={}`.

danielberndt replied:
Checked out [this part](https://emotion.sh/docs/css#css-prop) of the emotion docs!

`<span css={styles}/>` will be processed by the emotion babel plugin and turned into `<span className={...}/>`

TejasQ replied:
🤔 

Yes, but is it safe to assume that everyone has the babel plugin? I'd rather the codemod give users something that works with base emotion out of the box. 

We could even add a plugin option for users to opt-in to babel-plugin-specific syntax sugar like this. I know @kentcdodds [wanted this](https://github.com/paypal/glamorous/issues/419#issuecomment-397847519) and I agree that this would be a really nice feature to add to this codemod.

danielberndt replied:
So what would you suggest for code like:
```jsx
<glamorous.Span css={{":hover": {color: "red"}}}/>
``` 

I guess we could do something like:

```jsx
// opts.withPlugin === true
<span css={{":hover": {color: "red"}}}/>

// opts.withPlugin === false
import {css} from emotion
<span className={css({":hover": {color: "red"}})}/>

```

But for the sake of keeping this PR small, I'd suggest to assume that the plugin is installed for now. We can introduce the `withPlugin` flag in a separate PR

TejasQ replied:
Great suggestion.

Philosophically, for keeping the PR small, I'd remove the functionality around CSS props and other plugin-related transformations from _this_ PR altogether and include them all in a more encapsulated PR.

This means that the existing transformations around this functionality ought not be removed, but _moved_ to another PR so this one is a little leaner and simpler, even to review.

TejasQ commented on index.js:
```diff
@@ -13,133 +13,167 @@
+        );
+        newAttrs.push(cssAttr);
+      } else if (cssAttr.value.expression.type !== "ObjectExpression") {
+        // turn <span css={obj} .../> into <span css={{...obj}} .../>
```
I think the intent here is to turn `<span css={obj()} into <span css={{ ...obj() }}`? The current description of spreading an object into an object seems to be a bit of a no-op. 😅

danielberndt replied:
This code is needed to transform e.g.

```<gl.Span css={redStyle} marginTop={5}/>```

into

```<span css={{...redStyle, marginTop: 5}}/>```

So the comment doesn't tell the whole story yet, I'll fix it!

danielberndt commented:
> Also, I've noticed that changing the ImportDefaultSpecifier in the AST explorer outright removes the default import if it is not gl. 🤔

Ah, have a look at [this test](https://github.com/danielberndt/babel-plugin-glamorous-to-emotion/tree/a2ded59032180161f77d96724d8878ad7d47a637/__tests__/__fixtures__/02-jsx) to understand why I did this.

Previously a reference to glamorous was needed for `<glamorous.div/>`, but after the transform it's only `<div/>`. So we can remove the `import` statement, if nothing else refers to `glamorous`.

One side effect is that if glamorous was imported and never used, the import will disappear due to the codemod. That's something I'd be fine with.

danielberndt commented:
Okay, I just added the discussed improvements. (and added eslint)

If you're happy with the changes, I'd suggest to work on all the additional suggestions above (and in the code) in separate PRs :)

TejasQ commented:
Thanks for the documentation adjustments and destructured aliases! Would you agree with me that it is a _far_ more welcoming to readers of the source code now? I certainly think so. Great job!

I'll have another read through it tomorrow and possibly approve and merge it if we get [the separation of PR concerns](https://github.com/TejasQ/babel-plugin-glamorous-to-emotion/pull/2#discussion_r207722625) right.

One extra nit I found was that we're writing multiline comments with `//` on many lines. I've seen this pattern around the web and I have to say I don't get it. I _much_ prefer using the syntax we already have for multiline comments. But, as I said, it's just a nit and not at all pivotal to the way this tool works or even reads. 

It has been a real pleasure collaborating with you on this so far. 😄

danielberndt commented:
Re separating the css property stuff into a separate PR:
fine by me, but I'll add a warning to the user that the codemod is potentially outputting broken code if it encounters a `<glamorous.Div>` component. 
I also guess since a separate PR will be more focussed, so turnaround time will probably be quicker. (I still have two projects that I'd like to transform at some point soon)

Regarding multiline comments. To me it's mostly a matter of organical growth. Most comments start off as a single line comment, then I add a half-sentence, it gets too long, so I break it into two lines without changing the type of comment. It's a mixture of lazyness and lack of IDE-features I guess.

TejasQ commented on index.js:
```diff
@@ -13,133 +13,127 @@
-                        )
-                    );
-                }
+  const styledVisitor = {
```
Can we document this a little better?

`name`, `oldName`, etc. seem a little ambiguous.

TejasQ commented on package.json:
```diff
@@ -5,5 +5,27 @@
   "repository": "https://github.com/tejasq/babel-plugin-glamorous-to-emotion",
   "author": "Tejas Kumar <tejas@tejas.qa>",
-  "license": "MIT"
+  "license": "MIT",
```
You should add a `"collaborators"` here since we're buddies now! 😄

TejasQ commented:
Great stuff! Let's add a little more explicit naming and docs and then we're good to merge. 🚀 🔥

danielberndt commented:
Alright, I just added some clarifying comments for the core part of the transform step :)

TejasQ commented on index.js:
```diff
@@ -13,133 +13,136 @@
-                        ]),
+        // only if the traversal below wants to know the newName,
+        // we're gonna add the default import
+        const getNewName = () => {
```
How does this work on line 61 if it's defined here? 🤔 Surely I've missed something.

danielberndt replied:
Whatever you put as a second argument when calling a traverse step, will be available as `state` in the second parameter of each Visitor.

I pass the `getNewName` parameter [here](https://github.com/TejasQ/babel-plugin-glamorous-to-emotion/blob/127e82f102111122c3ebc685be389af73b5292ec/index.js#L127)

Check [this part of kentcdodds best practices](https://github.com/kentcdodds/babel-plugin-handbook/blob/master/README.md#optimizing-nested-visitors) for more information!

TejasQ replied:
Ah! I missed the part where it was being passed in on line 44! Makes sense!

danielberndt commented:
Ah sorry, didn't expect your review to be that quick, I just made a small change to how the "contributors" field is set up. And... this seems to have dismissed your review.
